profiles/killbill/src/main/java/org/killbill/billing/server/modules/JaxRSAopModule.java

```diff
@@ -0,0 +1,98 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.inject.AbstractModule;
+import com.google.inject.matcher.Matcher;
+import com.google.inject.matcher.Matchers;
+
+public class JaxRSAopModule extends AbstractModule {
+
+    private static final Logger logger = LoggerFactory.getLogger(KillbillApiAopModule.class);
+
+    private static final Matcher<Method> API_RESOURCE_METHOD_MATCHER = new Matcher<Method>() {
+        @Override
+        public boolean matches(final Method method) {
+            return !method.isSynthetic() &&
+                   (
+                           method.getAnnotation(DELETE.class) != null ||
+                           method.getAnnotation(GET.class) != null ||
+                           method.getAnnotation(HEAD.class) != null ||
+                           method.getAnnotation(OPTIONS.class) != null ||
+                           method.getAnnotation(POST.class) != null ||
+                           method.getAnnotation(PUT.class) != null
+                   );
+        }
+
+        @Override
+        public Matcher<Method> and(final Matcher<? super Method> other) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Matcher<Method> or(final Matcher<? super Method> other) {
+            throw new UnsupportedOperationException();
+        }
+    };
+
+    @Override
+    protected void configure() {
+        bindInterceptor(Matchers.subclassesOf(JaxrsResource.class),
+                        API_RESOURCE_METHOD_MATCHER,
+                        new JaxRsMethodInterceptor());
+    }
+
+    public static class JaxRsMethodInterceptor implements MethodInterceptor {
+
+        @Override
+        public Object invoke(final MethodInvocation invocation) throws Throwable {
+            return DBRouterUntyped.withRODBIAllowed(isRODBIAllowed(invocation),
+                                                    new WithProfilingCallback<Object, Throwable>() {
+                                                        @Override
+                                                        public Object execute() throws Throwable {
+                                                            logger.debug("Entering JAX-RS call {}, arguments: {}", invocation.getMethod(), invocation.getArguments());
+                                                            final Object proceed = invocation.proceed();
+                                                            logger.debug("Exiting  JXA-RS call {}, returning: {}", invocation.getMethod(), proceed);
+                                                            return proceed;
+                                                        }
+                                                    });
+        }
+
+        private boolean isRODBIAllowed(final MethodInvocation invocation) {
+            return invocation.getMethod().getAnnotation(GET.class) != null;
```

Comment:
Questions/issues ?

* Are we really guaranteed that `GET` operation never end up writing state ? For instance, in a path of retrieving payments, could not we end up calling Janitor logic, then completing payment ?
* What would happen if a plugin gets invoked during get -- e.g PaymentApiType.NOTIFICATION. What would then happen to this thread ?

Reply:
Also, conversely, less important, but calls like dryRun invoice, currently implemented using `POST` would not benefit from this logic.

Reply:
> * Are we really guaranteed that `GET` operation never end up writing state ? For instance, in a path of retrieving payments, could not we end up calling Janitor logic, then completing payment ?

Nope, we aren't. That's why I have this check in the lower levels: https://github.com/killbill/killbill/pull/962/files#diff-cc1298714ab6fed9365a7fab751f800cR72.

See test: https://github.com/killbill/killbill/pull/962/files#diff-3aa8c9662fc08477a4d50d574aba0408R83.

> * What would happen if a plugin gets invoked during get -- e.g PaymentApiType.NOTIFICATION. What would then happen to this thread ?

When is `PaymentApiType.NOTIFICATION` used?

Regarding plugins called synchronously, we switch to a different thread pool (`pay-plugin-grp`) and the state isn't carried over (so it would be `RW_ONLY` by default, unless a `ROTenantContext` is used).

> Also, conversely, less important, but calls like dryRun invoice, currently implemented using `POST` would not benefit from this logic.

Indeed, that's one drawback of the new implementation (simpler, but less optimized in some edge cases).

Reply:
> Regarding plugins called synchronously, we switch to a different thread pool (`pay-plugin-grp`) and the state isn't carried over (so it would be `RW_ONLY` by default, unless a `ROTenantContext` is used).

We actually don't switch for `GET`, only for RW calls. Behavior tested with 7aa63b274a1e458cca5991c618e214cb8c742668.